Source/WebCore/PAL/pal/cocoa/VImageConverterPtr.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <Accelerate/Accelerate.h>
+#include <wtf/RawPtrTraits.h>
+#include <wtf/RefPtr.h>
+
+struct VImageConverterRefDerefTraits {
+    static ALWAYS_INLINE void refIfNotNull(vImageConverterRef converter)
+    {
+        if (LIKELY(converter != nullptr))
+            vImageConverter_Retain(converter);
+    }
+
+    static ALWAYS_INLINE void derefIfNotNull(vImageConverterRef converter)
+    {
+        if (LIKELY(converter != nullptr))
+            vImageConverter_Release(converter);
+    }
+};
+
+using VImageConverterPtr = RefPtr<vImageConverter, WTF::RawPtrTraits<vImageConverter>, VImageConverterRefDerefTraits>;
+
+inline VImageConverterPtr adoptVImageConverter(vImageConverterRef converter)
+{
+    return adoptRef<vImageConverter, WTF::RawPtrTraits<vImageConverter>, VImageConverterRefDerefTraits>(converter);
+}
```

Comment:
There seems to be two different ways of dealing with `RefPtr` traits in the codebase.

One is like this, where we define a new type alias for refcounted pointers of this specific type, so that users do not need to type out the full `RefPtr<...>` with the traits type argument. A new `adoptXxx` function is needed for this type.

The other is to specialize `DefaultRefDerefTraits`, which I see done for some Gtk types (in RefPtrCairo.h and RefPtrFontconfig.h). With this, users can write `RefPtr<TheType>` and use `adoptRef`.

Is there a reason to choose one way over the other? We should probably have some guidance in Introduction.md in the `RefPtr` section. (Not something to do in this patch.)

----

But specifically for `vImageConverterPtr`: is this needed, or does `RetainPtr` work on it? I see in PixelBufferConversion.cpp that we use `adoptCF` on the return value of `vImageConverter_CreateWithCGImageFormat`.

Reply:
I started a thread in internal slack about this.